provider/datagen/src/transform/collator/transform.rs

```
@@ -0,0 +1,339 @@
+                    Ok(DataResponse {
+                        metadata: DataResponseMetadata::default(),
+                        // The struct conversion is macro-based instead of
+                        // using a method on the Serde struct, because the
+                        // method approach caused lifetime issues that I
+                        // didn't know how to solve.
+                        payload: Some(DataPayload::from_owned($conversion)),
+                    })
+                } else {
+                    Err(DataErrorKind::MissingResourceOptions.with_req($marker::KEY, req))
+                }
+            }
+        }
+
+        icu_provider::impl_dyn_provider!(
+            $provider,
+            [$marker,],
+            SERDE_SE,
+            ITERABLE_SERDE_SE,
+            DATA_CONVERTER
+        );
+
+        impl IterableResourceProvider<$marker> for $provider {
+            fn supported_options(
+                &self,
+            ) -> Result<Box<dyn Iterator<Item = ResourceOptions>>, DataError> {
+                if self.data.read().unwrap().is_none() {
+                    let root_dir = self.source.get_coll_root()?;
+                    let data = $provider::load_data(root_dir).map_err(|e| {
+                        DataError::custom("Could not create provider").with_display_context(&e)
+                    })?;
+                    *self.data.write().unwrap() = Some(data);
+                }
+
+                let guard = self.data.read().unwrap();
+
+                let list: Vec<ResourceOptions> = guard
+                    .as_ref()
+                    .unwrap()
+                    .keys()
+                    .map(|k| {
+                        let (language, variant) = k.rsplit_once('_').unwrap();
+                        let langid = if language == "root" {
+                            LanguageIdentifier::default()
+                        } else {
+                            language.parse().unwrap()
```

Nit: use `filter_map` to exclude ignore invalid data instead of panicking.

---

Done even though I think it's unhelpful to silently ignore badly-named files instead of alerting the user of datagen.

---

In that case return a `DataError`, but don't panic